diff --git a/constructs/NextjsSite.js b/constructs/NextjsSite.js
index 36465764aa6aca0f30f52c67b6c757f9dd95316f..b6574ab7a8ad1c0f430c34829f6e774fac51e0d1 100644
--- a/constructs/NextjsSite.js
+++ b/constructs/NextjsSite.js
@@ -1,12 +1,17 @@
 import fs from "fs";
 import path from "path";
-import { Duration as CdkDuration } from "aws-cdk-lib/core";
+import { Duration as CdkDuration, CustomResource } from "aws-cdk-lib/core";
 import { Code, Runtime, Function as CdkFunction, Architecture, } from "aws-cdk-lib/aws-lambda";
 import { Queue } from "aws-cdk-lib/aws-sqs";
 import { SqsEventSource } from "aws-cdk-lib/aws-lambda-event-sources";
 import { Stack } from "./Stack.js";
 import { SsrSite } from "./SsrSite.js";
 import { toCdkSize } from "./util/size.js";
+import { AttributeType, BillingMode, Table } from "aws-cdk-lib/aws-dynamodb";
+import { Provider } from "aws-cdk-lib/custom-resources";
+import { PolicyStatement } from "aws-cdk-lib/aws-iam";
+import { RetentionDays } from "aws-cdk-lib/aws-logs";
+
 /**
  * The `NextjsSite` construct is a higher level CDK construct that makes it easy to create a Next.js app.
  * @example
@@ -194,6 +199,59 @@ export class NextjsSite extends SsrSite {
         server?.addEnvironment("REVALIDATION_QUEUE_URL", queue.queueUrl);
         server?.addEnvironment("REVALIDATION_QUEUE_REGION", Stack.of(this).region);
         queue.grantSendMessages(server?.role);
+        const app = this.node.root;
+
+        const table = new Table(this, "RevalidationTable", {
+            partitionKey: { name: "tag", type: AttributeType.STRING },
+            sortKey: { name: "path", type: AttributeType.STRING },
+            pointInTimeRecovery: true,
+            billingMode: BillingMode.PAY_PER_REQUEST,
+            tableName: app.logicalPrefixedName("next-revalidation-table"),
+        });
+
+        table.addGlobalSecondaryIndex({
+            indexName: "revalidate",
+            partitionKey: { name: "path", type: AttributeType.STRING },
+            sortKey: { name: "revalidatedAt", type: AttributeType.NUMBER },
+        });
+
+        server?.addEnvironment("CACHE_DYNAMO_TABLE", table.tableName);
+        table.grantReadWriteData(server?.role);
+
+        const insertFn = new CdkFunction(this, "RevalidationInsertFunction", {
+            description: "Next.js revalidation data insert",
+            handler: "index.handler",
+            code: Code.fromAsset(
+                path.join(this.props.path, ".open-next", "dynamodb-provider")
+            ),
+            runtime: Runtime.NODEJS_18_X,
+            timeout: CdkDuration.minutes(14),
+            initialPolicy: [
+                new PolicyStatement({
+                    actions: [
+                        "dynamodb:BatchWriteItem",
+                        "dynamodb:PutItem",
+                        "dynamodb:DescribeTable",
+                    ],
+                    resources: [table.tableArn],
+                }),
+            ],
+            environment: {
+                CACHE_DYNAMO_TABLE: table.tableName,
+            }
+        });
+
+        const provider = new Provider(this, "RevalidationProvider", {
+            onEventHandler: insertFn,
+            logRetention: RetentionDays.ONE_DAY,
+        });
+
+        new CustomResource(this, "RevalidationResource", {
+            serviceToken: provider.serviceToken,
+            properties: {
+                version: Date.now().toString()
+            }
+        });
     }
     getConstructMetadata() {
         return {
@@ -202,3 +260,4 @@ export class NextjsSite extends SsrSite {
         };
     }
 }
+
